improved doc rhat

DavAug opened the pull request:
Closes #1144 
Closes #1158 
Closes #1157 

I extended the doc string for Rhat according to the reference. I hope it makes it clearer what happens.

martinjrobins commented on pints/_diagnostics.py:
```diff
@@ -100,15 +100,56 @@ def reorder_all_params(chains):
+    :math:`\hat{R}` diagnoses convergence by checking mixing and stationarity
+    of :math:`M` chains (at least two, :math:`M\geq 2`). To diminish the
+    influence of starting values, the first half of each chain is considered
+    as warm up and does not enter the calculation. Subsequently, the truncated
```
looking at the code, it looks like it doesn't discard warm-up. Did I miss something?

DavAug commented:
That is correct. It also doesn’t split the chains into 2. But the previous doc had said that it computes the Rhat according to the reference and that is what the reference says. This is why we had the discussion in the meeting just now.

I will update the code in a bit according to the definition, while introducing an argument warm_up which determines which fraction is supposed to be thrown away and set it’s default to 0.

On Jun 2, 2020, at 1:33 PM, Martin Robinson <notifications@github.com<mailto:notifications@github.com>> wrote:



@martinjrobins commented on this pull request.

________________________________

In pints/_diagnostics.py<https://github.com/pints-team/pints/pull/1156#discussion_r433808363>:

>      r"""
-    Calculates :math:`\hat{R} = \sqrt{(W(n - 1) / n + (1 / n) B) / W}` as per
-    [1]_ for a single parameter. It does this after splitting each chain into
-    two.
-    """
-    W = within(samples)
-    B = between(samples)
-    t = len(samples[0])
+    Returns the convergence measure :math:`\hat{R}` according to [1]_.
+
+    :math:`\hat{R}` diagnoses convergence by checking mixing and stationarity
+    of :math:`M` chains (at least two, :math:`M\geq 2`). To diminish the
+    influence of starting values, the first half of each chain is considered
+    as warm up and does not enter the calculation. Subsequently, the truncated


looking at the code, it looks like it doesn't discard warm-up. Did I miss something?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pints-team/pints/pull/1156#pullrequestreview-422587773>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEY2T3UUUODPTRZSCP5AN7LRUTPSHANCNFSM4NQRVKFQ>.

DavAug commented:
@MichaelClerx @ben18785 @martinjrobins @chonlei 

I refactored the rhat computation a little bit and adjusted the documentation. 
I used some numpy magic to combine `pints.rhat()` and `pints.rhat_all_params` into a single function `pints.rhat()`. This function takes multiple chains in form of numpy.ndarray as before, but now both dimensions (m, n) and (m, n, p) are accepted, where m is the number of chains, n is the number of samples per chain, and p is the number of parameters.

DavAug commented:
I also added some asserts which should be satisfied for the rhat computation to make sense. One of the assertions "the number of chains has to be at least 2, m>=2" breaks a test for the `pints.summary()`, because here only a single chain is passed to rhat. I am tempted to adjust the test as mixing of chains can only be tested if you have at least 2. 

Or is the rhat computation for a single chain useful in certain circumstances? In that case we could just remove the assert.

What do you think?

ben18785 commented on docs/source/diagnostics.rst:
```diff
@@ -9,14 +9,11 @@ Pints provides a number of functions to diagnose MCMC progress and convergence.
 
 - :func:`rhat`
-- :func:`rhat_all_params`
 - :func:`effective_sample_size`
```
Sorry, realise they weren't before, but can we make these alphabetical?

DavAug replied:
Done

ben18785 commented on examples/sampling/adaptive-covariance-haario.ipynb:
```diff
@@ -139,7 +139,7 @@
     "# Check convergence using rhat criterion\n",
     "print('R-hat:')\n",
-    "print(pints.rhat_all_params(chains))\n",
+    "print(pints.rhat(chains))\n",
```
Thanks for changing these: what'd I'd prefer though would be to change all of these over to using the sampler results object (the one that prints a table of parameter quantiles). I think these are much neater and display everything that's needed. See the ``adaptive-covariance-haario-bardenet.ipynb`` for an example of this.

ben18785 commented on examples/sampling/adaptive-covariance-rao-blackwell.ipynb:
```diff
@@ -138,7 +138,7 @@
     "# Check convergence using rhat criterion\n",
     "print('R-hat:')\n",
-    "print(pints.rhat_all_params(chains))\n",
+    "print(pints.rhat(chains))\n",
```
Same as above. For all MCMC notebooks, I think this should be the case.

ben18785 commented on pints/_diagnostics.py:
```diff
@@ -55,72 +55,185 @@ def effective_sample_size(samples):
 
     References
     ----------
     ..  [1] "Bayesian data analysis", 3rd edition, Gelman et al., 2014.
```
Can we point to the relevant chapter in this? Think it's chapter 11. Also, I think we should cite Stan here, since the splitting of chains in two is what they did (I don't think it's mentioned in Gelman).

DavAug replied:
Yes chapter 11, page 285 equation 11.4. I implemented the method strictly according to this definition. Do you maybe mean the rank-based computation that Stan does, and is not mentioned in this book?

ben18785 replied:
Nah, I was just meaning about splitting the chains in two. Does Gelman suggest this? If so, fine to keep as is. I think we can save doing any rank-based stuff for now.

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        # warm-up greater than 100%
+        warm_up = 1.1
+        message = (
+            'warum_up is set to 1.1. warm_up only takes values in [0,1].')
```
''warm-up''

DavAug replied:
I referring here to the input variable `warm_p`. I would therefore leave it like that. But I can add some `.` marks if that make sit clearer.

ben18785 replied:
My comment was just about spelling warm_up rather than warum_up

DavAug replied:
Ahh I see :D

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        self.assertRaisesRegex(
+            ValueError, message[0], pints.rhat, chains, warm_up)
 
-        x = np.array([[1.0, 1.1, 1.4, 1.3, 1.3],
```
Is the reason this test has a different result this time because the chains are being split? As in Rhat=1.43115 before, now it's 2.33...

DavAug replied:
Yes the old rhat did not test for stationarity. This rhat always splits the chains in half and that's why the old and new rhat do not agree. But I implemented the methods to work for one parameter only, and afterwards realised that broadcasting also allows for the computation of multiple parameters with the same expressions. And for multiple params the rhat agrees with the old results. So I'm pretty confident that rhat is correct.

ben18785 replied:
Ok, cool

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        chains = np.empty(shape=1)
+        message = (
+            'Dimension of chains is 1. '
+            + 'Method computes R^hat for one '
```
Rhat rather than R^hat

DavAug replied:
Done

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        # Pass chain of dimension 1
+        chains = np.empty(shape=1)
+        message = (
+            'Dimension of chains is 1. '
```
Can I just check -- does this method still work for models with only a single parameter?

DavAug replied:
Yes it works for input dimensions (m, n) and (m, n, p) where m is the number of chains, n the number of samples per chain and p the number of parameters.

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        chains = np.empty(shape=(1, 1, 1, 1))
+        message = (
+            'Dimension of chains is 4. '
+            + 'Method computes R^hat for one '
```
Rhat

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        chains = np.array([[1.0, 1.1, 1.4, 1.3],
+                          [1.0, 2.0, 3.0, 4.0]])
+        self.assertAlmostEqual(
+            pints._diagnostics.rhat(chains), 2.3303847470550716, 6)
```
This looks like a new value of Rhat versus the old test value...

DavAug replied:
This is because the old Rhat did not split the chains in half.

ben18785 replied:
Cool

ben18785 commented on pints/tests/test_diagnostics.py:
```diff
@@ -79,36 +79,99 @@ def test_within(self):
+        # Negative warm-up
+        warm_up = -0.1
+        message = (
+            'warum_up is set to -0.1. warm_up only takes values in [0,1].')
```
warm-up

ben18785 replied:
Again, just warum_up vs warm_up
